libminifi/include/core/repository/AtomicRepoEntries.h

```diff
@@ -0,0 +1,501 @@
+      return;
+    }
+    if (ref_count_ > 0){
+      ref_count_--;
+    }
+    if (ref_count_ == 0 && free_required)
+    {
+      size_t bufferSize = value_.getBufferSize();
+      value_.clearBuffer();
+      has_value_ = false;
+      if (total_size_ != nullptr) {
+	*total_size_ -= bufferSize;
+      }
+      free_required = false;
+    }
+    else{
+    }
+    try_unlock();
+  }
+
+  /**
+   * Moved the value into the argument
+   * @param value the previous value will be moved into this parameter
+   * @return  success of get operation based on whether or not this atomic entry has a value.
+   */
+  bool getValue(const T &key, RepoValue<T> **value) {
+    try_lock();
+    if (!has_value_) {
+      try_unlock();
+      return false;
+    }
+    if (!value_.isKey(key)) {
+      try_unlock();
+      return false;
+    }
+    ref_count_++;
+    *value = &value_;
+    try_unlock();
+    return true;
+  }
+
+  /**
+   * Operation that will be used to test and free if a release is required without
+   * setting a new object.
+   * @param releaseTest function that will be used to free the RepoValue key from
+   * this atomic entry.
+   * @param freedValue informs the caller if an item was freed.
+   */
+  T testAndFree(std::function<bool(T)> releaseTest, bool &freedValue) {
+    try_lock();
+    T ref;
+    if (!has_value_) {
+      try_unlock();
+      return ref;
+    }
+
+    if (releaseTest(value_.getKey())) {
+      size_t bufferSize = value_.getBufferSize();
+      value_.clearBuffer();
+      ref = value_.getKey();
+      has_value_ = false;
+      if (total_size_ != nullptr) {
+        *total_size_ -= bufferSize;
+      }
+
+    }
+    try_unlock();
+    return ref;
+  }
+  
+  size_t getLength()
+  {
+    size_t size = 0;
+     try_lock();
+     size = value_.getBufferSize();
+     try_unlock();
+     return size;
+     
+  }
+
+  /**
+   * sets has_value to false; however, does not call
+   * any external entity to further free RepoValue
+   */
+  bool freeValue(const T &key) {
+    try_lock();
+    if (!has_value_) {
+      try_unlock();
+      return false;
+    }
+    if (!value_.isKey(key)) {
+      try_unlock();
+      return false;
+    }
+    if (ref_count_ > 0)
+    {
+       free_required = true;
+       try_unlock();
+       return true;
+    }
+    size_t bufferSize = value_.getBufferSize();
+    value_.clearBuffer();
+    has_value_ = false;
+    if (total_size_ != nullptr) {
+      *total_size_ -= bufferSize;
+    }
+    free_required = false;
+    try_unlock();
+    return true;
+  }
+
+  /**
+   * Appends buffer onto this atomic entry if key matches
+   * the current RepoValue's key.
+   */
+  bool insert(const T key, uint8_t *buffer, size_t size) {
+    try_lock();
+
+    if (!has_value_) {
+      try_unlock();
+      return false;
+    }
+
+    if (!value_.isKey(key)) {
+      try_unlock();
+      return false;
+    }
+
+    if ((total_size_ != nullptr && max_size_ != nullptr) && (*total_size_ + size > *max_size_)) {
```

Comment:
These are pointers to a global max size for that volatile repo? Is it confusing? I can rename it to make it clearer.